### PR TITLE
The Three "R"s: Rust, REPL, Runner

### PR DESCRIPTION
**Pull request by ozbe**
REPLs and runners can be a quick and easy way to iterate and execute code. Let's try different Rust REPLs & runners to see the differences between each tool.

/schedule 2021-01-21

### CODE REVIEW

File: content/posts/the-three-rs-rust-repl-and-runners/index.mdx [hunk 1]
```diff
@@ -0,0 +1,234 @@
+---
+title: 'The Three "R"s: Rust, REPL, Runner'
+date: 2021-01-22
+---
+
+REPLs and runners can be an easier and quick way to iterate and execute code. Apply the same Rust example with a Cargo package and different Rust REPLS & Runners to see the differences between each tool.
```
Comment (wezm) [inline content/posts/the-three-rs-rust-repl-and-runners/index.mdx:6]: This reads a bit awkwardly to me. Perhaps:

> REPLs and runners can be a quick and easy way to iterate and execute code.

I'm not quite following the second sentence so not sure what to suggest there.

File: content/posts/the-three-rs-rust-repl-and-runners/index.mdx [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+	2. No, create a new Cargo package
+2. Add the dependency to `Cargo.toml`
+3. Change some relevant code to try something out about the library
+4. Build
+5. Run
+6. Do I have everything I need to know to make a decision about the library
+	1. Yes, go do what I need to do
+	2. No, go back to Step 3
+
+These steps can grow tedious, especially when making a new cargo package, and the iterations can be shortened by more work up front looking at documentation, source, and examples. But sometimes you just want to experiment in a quick and efficient way. That is when REPLs and runners can prove to be helpful.
```
Comment (wezm) [inline content/posts/the-three-rs-rust-repl-and-runners/index.mdx:23]:
```suggestion
These steps can grow tedious, especially when making a new cargo package. The iterations can be shortened by more work up front looking at documentation, source, and examples. But sometimes you just want to experiment in a quick and efficient way. That is when REPLs and runners can prove to be helpful.
```

File: content/posts/the-three-rs-rust-repl-and-runners/index.mdx [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+
+For the uninitiated, a REPL is typically an interactive shell that takes user input, evaluates said input, and returns the result.  Assuming I can access the required dependencies, I should be able to take the contents of the `fn main()` above and enter each expressions in line-by-line in a REPL.
```
Comment (wezm) [inline content/posts/the-three-rs-rust-repl-and-runners/index.mdx:25]:
```suggestion
For the uninitiated, a REPL (Read, Evaluate, Print, Loop) is typically an interactive shell that takes user input, evaluates said input, and returns the result.  Assuming I can access the required dependencies, I should be able to take the contents of the `fn main()` above and enter each expressions in line-by-line into a REPL.
```

File: content/posts/the-three-rs-rust-repl-and-runners/index.mdx [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+
+I did some digging to find an official Rust REPL and there is an open issue to officially support [implement a proper REPL · Issue #655 · rust-lang/rfcs · GitHub](https://github.com/rust-lang/rfcs/issues/655).
```
Comment (wezm) [inline content/posts/the-three-rs-rust-repl-and-runners/index.mdx:27]:
```suggestion
I did some digging to find an official Rust REPL and there is an open issue to officially [implement a proper REPL](https://github.com/rust-lang/rfcs/issues/655).
```

File: content/posts/the-three-rs-rust-repl-and-runners/index.mdx [hunk 5]
```diff
@@ -0,0 +1,234 @@
...
+
+While there isn’t an official REPL, there is at least one active Rust REPL:
+* [Evcxr REPL](https://github.com/google/evcxr/blob/master/evcxr_repl/)
+
+A Rust runner allows you to run (well, compile, link, _and_ run) Rust without the _bloat_ of Cargo. Similar to the REPLs, there isn’t an official Rust runner, besides Cargo, and there are a few options:
```
Comment (wezm) [inline content/posts/the-three-rs-rust-repl-and-runners/index.mdx:32]: Is cargo really bloat if it's part of the a standard Rust install so you can rely on anyone coding in Rust having it? Or do you mean that without the hassle of creating a full cargo project and the structure it entails. If it's the latter then I'd suggest rewording to make it clearer.

File: content/posts/the-three-rs-rust-repl-and-runners/index.mdx [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+$ rustup component add rust-src
+```
+
+Now you can go ahead and install the binary:
+```bash
+$ cargo install evcxr_repl
+```
+
+### Run
+Enter the following lines in (without the `>> `) the REPL to get the _same_ output as the Cargo package :
```
Comment (wezm) [inline content/posts/the-three-rs-rust-repl-and-runners/index.mdx:102]:
```suggestion
Enter the following lines in the REPL (without the `>> `) to get the _same_ output as the Cargo package :
```

File: content/posts/the-three-rs-rust-repl-and-runners/index.mdx [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+}
+>> let point = Point { x: 1, y: 2 };
+>> let serialized = serde_json::to_string(&point).unwrap();
+>> println!("serialized = {}", serialized);
+>> let deserialized: Point = serde_json::from_str(&serialized).unwrap();
+>> println!("deserialized = {:?}", deserialized);
+```
+
+A few things to note about the REPL session:
+1. `:dep` lines are equivalent to the `[dependencies` section of the `Cargo.toml` in the Cargo package
```
Comment (wezm) [inline content/posts/the-three-rs-rust-repl-and-runners/index.mdx:121]:
```suggestion
1. `:dep` lines are equivalent to the `[dependencies]` section of the `Cargo.toml` in the Cargo package
```

**Review by wezm — COMMENTED** [review]
Saw your Meetup post so thought I leave some comments. Overall I think the post content is shaping up well. I was aware of some of these tools but had never tried them, so this post definitely taught me some things.

File: content/posts/the-three-rs-rust-repl-and-runners/index.mdx (hunk 1 above)
Comment (ozbe) [reply content/posts/the-three-rs-rust-repl-and-runners/index.mdx:6]: Your suggestion reads much better to me as well. I added a better segue to a slightly altered version of the second sentence.
```
--- a/content/posts/the-three-rs-rust-repl-and-runners/index.mdx
+++ b/content/posts/the-three-rs-rust-repl-and-runners/index.mdx
@@ -3,7 +3,7 @@ title: 'The Three "R"s: Rust, REPL, Runner'
 date: 2021-01-22
 ---
 
-REPLs and runners can be an easier and quick way to iterate and execute code. Apply the same Rust example with a Cargo package and different Rust REPLS & Runners to see the differences between each tool.
+REPLs and runners can be a quick and easy way to iterate and execute code. Let's try different Rust REPLs & runners to see the differences between each tool.
 
 ## Intro
 tl;dr [Jump to the code](#cargo-project)
```

File: content/posts/the-three-rs-rust-repl-and-runners/index.mdx (hunk 5 above)
Comment (ozbe) [reply content/posts/the-three-rs-rust-repl-and-runners/index.mdx:32]: It is certainly the latter. Seems I may have left a bit too much masked behind `_bloat_`. I've changed the language to better reflect what I meant.

```
--- a/content/posts/the-three-rs-rust-repl-and-runners/index.mdx
+++ b/content/posts/the-three-rs-rust-repl-and-runners/index.mdx
@@ -29,7 +29,7 @@ I did some digging to find an official Rust REPL and there is an open issue to o
 While there isn’t an official REPL, there is at least one active Rust REPL:
 * [Evcxr REPL](https://github.com/google/evcxr/blob/master/evcxr_repl/)
 
-A Rust runner allows you to run (well, compile, link, _and_ run) Rust without the _bloat_ of Cargo. Similar to the REPLs, there isn’t an official Rust runner, besides Cargo, and there are a few options:
+A Rust runner allows you to run (well, compile, link, _and_ run) Rust code without the overhead (and benefits) of creating a Cargo package. Similar to the REPLs, there isn’t an official Rust runner, besides Cargo, and there are a few options:
 * [runner](https://github.com/stevedonovan/runner)
 * [cargo-eval](https://github.com/reitermarkus/cargo-eval)
 ```

**Review by ozbe — COMMENTED** [review]
> Saw your Meetup post so thought I leave some comments. Overall I think the post content is shaping up well. I was aware of some of these tools but had never tried them, so this post definitely taught me some things.

@wezm, I appreciate the feedback. I've applied some of your suggestions and made alterations to try to address your comments. Glad you learned a few things and hope some others can find value in this post as well. Thanks for helping improve it.
